### PR TITLE
Load a set of predefined labels

### PR DESCRIPTION
**Pull request by 0xbaadf00d**
This pull request add ability to load a set of predefined labels (Issue #1562). This feature is really useful for user who create same labels on a lot of repositories.

__Default file__

The default set of predefined labels only contains:

```
#e11d21 bug
#009800 enhancement
#eb6420 task
#fbca04 suggestion
#000000 won't fix
```

__Customization:__

User can customize available label sets by adding files on `custom/conf/label` directory. The file must respect the following format : _&lt;html-color>&lt;space>&lt;label-name>_

__Preview:__

In the following screenshots, _Github_ and _Gitlab_ presets are not included on this pull request. Their two files are only on my own `custom/conf/label` directory.

labeltemplate-0
labeltemplate-1
labeltemplate-2

### CODE REVIEW

**Comment by tgy** [comment]
Thanks :)

**Comment by Unknwon** [comment]
Seems having conflict..

**Comment by 0xbaadf00d** [comment]
I fix this today

**Comment by Unknwon** [comment]
doesn't seem fixed...?

**Comment by 0xbaadf00d** [comment]
Its due to some files modified by previously accepted merge requests since yesterday. Fix push.

File: cmd/web.go [hunk 1]
```diff
@@ -477,6 +477,7 @@ func runWeb(ctx *cli.Context) error {
 			m.Post("/new", bindIgnErr(auth.CreateLabelForm{}), repo.NewLabel)
 			m.Post("/edit", bindIgnErr(auth.CreateLabelForm{}), repo.UpdateLabel)
 			m.Post("/delete", repo.DeleteLabel)
+			m.Post("/initialize", bindIgnErr(auth.InitializeLabelsForm{}), repo.InitializeLabelsFromTemplate)
```
Comment (Unknwon) [inline cmd/web.go]: `InitializeLabelsFromTemplate` -> `InitializeLabels`

File: conf/label/Default [hunk 2]
```diff
@@ -0,0 +1,5 @@
+#e11d21 bug
```
Comment (Unknwon) [inline conf/label/Default]: I think default to be same as GitHub is better. Ref: https://github.com/unknwon/sgtest/labels

File: conf/locale/locale_en-US.ini [hunk 3]
```diff
@@ -492,6 +492,10 @@ issues.create = Create Issue
 issues.new_label = New Label
 issues.new_label_placeholder = Label name...
 issues.create_label = Create Label
+issues.labeltemplates.title=Load a set of labels
```
Comment (Unknwon) [inline conf/locale/locale_en-US.ini]: `labeltemplates` -> `label_templates`

File: modules/auth/repo_form.go [hunk 4]
```diff
@@ -216,10 +216,18 @@ type CreateLabelForm struct {
 	Color string `binding:"Required;Size(7)" locale:"repo.issues.label_color"`
 }
 
+type InitializeLabelsForm struct {
+	TemplateName string `form:"templatename" binding:"Required"`
```
Comment (Unknwon) [inline modules/auth/repo_form.go]: Remove `form:"templatename"`

File: modules/auth/repo_form.go [hunk 5]
```diff
@@ -216,10 +216,18 @@ type CreateLabelForm struct {
...
+}
+
 func (f *CreateLabelForm) Validate(ctx *macaron.Context, errs binding.Errors) binding.Errors {
 	return validate(errs, ctx.Data, f, ctx.Locale)
 }
 
+func (f *InitializeLabelsForm) Validate(ctx *macaron.Context, errs binding.Errors) binding.Errors {
```
Comment (Unknwon) [inline modules/auth/repo_form.go]: Don't mix up code.

File: routers/repo/issue.go [hunk 6]
```diff
@@ -20,10 +20,13 @@ import (
 	"github.com/gogits/gogs/models"
 	"github.com/gogits/gogs/modules/auth"
 	"github.com/gogits/gogs/modules/base"
+	"github.com/gogits/gogs/modules/bindata"
 	"github.com/gogits/gogs/modules/context"
 	"github.com/gogits/gogs/modules/log"
 	"github.com/gogits/gogs/modules/markdown"
 	"github.com/gogits/gogs/modules/setting"
+	"path"
+	"regexp"
```
Comment (Unknwon) [inline routers/repo/issue.go]: https://github.com/Unknwon/go-code-convention/blob/master/en-US/import_packages.md

File: modules/auth/repo_form.go [hunk 7]
```diff
@@ -220,6 +220,16 @@ func (f *CreateLabelForm) Validate(ctx *macaron.Context, errs binding.Errors) bi
 	return validate(errs, ctx.Data, f, ctx.Locale)
 }
 
+// Label templates
+
+type InitializeLabelsForm struct {
+	TemplateName string `form:"templatename" binding:"Required"`
```
Comment (0xbaadf00d) [inline modules/auth/repo_form.go]: I dont know why, but without `form:"templatename"` I can't retrieve value from HTML form and I get the error `TemplateName cannot be empty.`

image

File: templates/repo/issue/labels.tmpl [hunk 8]
```diff
@@ -36,35 +36,34 @@
...
-							<p>{{.i18n.Tr "repo.issues.labeltemplates.info"}}</p>
+							<p>{{.i18n.Tr "repo.issues.label_templates.info"}}</p>
 							<br/>
 							<form class="ui form center" action="{{.Link}}/initialize" method="post">
 								{{.CsrfTokenHtml}}
 								<div class="field">
 									<div class="ui selection dropdown">
-										<input type="hidden" name="templatename" value="Default">
-										<div class="default text">{{.i18n.Tr "repo.issues.labeltemplates.helper"}}</div>
+										<input type="hidden" name="templatename" id="templatename" value="Default">
```
Comment (Unknwon) [inline templates/repo/issue/labels.tmpl]: `templatename`-> `template_name`

File: modules/auth/repo_form.go (hunk 7 above)
Comment (Unknwon) [inline modules/auth/repo_form.go]: See comment: https://github.com/gogits/gogs/pull/3459/files/e7a4b52213c19c0da12a88a0d7328f32b77d5f5d..63a948003a83d8dd545243c71be52aabaf026818#r76531134

File: templates/repo/issue/labels.tmpl (hunk 8 above)
Comment (0xbaadf00d) [inline templates/repo/issue/labels.tmpl]: yes, thanks you

**Comment by Unknwon** [comment]
Hmm, I'm sorry but we have conflict again..

**Comment by Unknwon** [comment]
Thanks!

**Pull request merged**